vpr/src/route/bucket.cpp

```diff
@@ -0,0 +1,234 @@
+#include "bucket.h"
+
+#include <cmath>
+#include "vtr_log.h"
+
+BucketItems::BucketItems() noexcept
+    : alloced_items_(0)
+    , num_heap_allocated_(0)
+    , heap_free_head_(nullptr) {}
+
+Bucket::Bucket() noexcept
+    : seed_(1231)
+    , heap_(nullptr)
+    , heap_size_(0)
+    , heap_head_(std::numeric_limits<size_t>::max())
+    , heap_tail_(0)
+    , conv_factor_(0.f)
+    , min_cost_(0.f)
+    , max_cost_(0.f) {}
+
+Bucket::~Bucket() {
+    free_all_memory();
+}
+
+void Bucket::init_heap(const DeviceGrid& grid) {
+    vtr::free(heap_);
+    heap_ = nullptr;
+
+    heap_size_ = (grid.width() - 1) * (grid.height() - 1);
+    heap_ = (BucketItem**)vtr::malloc(heap_size_ * sizeof(BucketItem*));
+    memset(heap_, 0, heap_size_ * sizeof(t_heap*));
+
+    heap_head_ = std::numeric_limits<size_t>::max();
+    heap_tail_ = 0;
+
+    conv_factor_ = kDefaultConvFactor;
+
+    min_cost_ = std::numeric_limits<float>::max();
+    max_cost_ = std::numeric_limits<float>::min();
+}
+
+void Bucket::free_all_memory() {
+    vtr::free(heap_);
+    heap_ = nullptr;
+
+    items_.free();
+}
+
+void Bucket::expand(size_t required_number_of_buckets) {
+    auto old_size = heap_size_;
+    heap_size_ = required_number_of_buckets * 2;
+
+    heap_ = (BucketItem**)vtr::realloc((void*)(heap_),
+                                       heap_size_ * sizeof(BucketItem*));
+    std::fill(heap_ + old_size, heap_ + heap_size_, nullptr);
+}
+
+void Bucket::verify() {
+    for (size_t bucket = heap_head_; bucket <= heap_tail_; ++bucket) {
+        for (BucketItem* data = heap_[bucket]; data != nullptr;
+             data = data->next_bucket) {
+            VTR_ASSERT(data->item.cost > 0 && ((size_t)cost_to_int(data->item.cost)) == bucket);
+        }
+    }
+}
+
+void Bucket::empty_heap() {
+    if (heap_head_ != std::numeric_limits<size_t>::max()) {
+        std::fill(heap_ + heap_head_, heap_ + heap_tail_ + 1, nullptr);
+    }
+    heap_head_ = std::numeric_limits<size_t>::max();
+    heap_tail_ = 0;
+
+    // Quickly reset all items to being free'd
+    items_.clear();
+}
+
+void Bucket::check_scaling() {
+    float min_cost = min_cost_;
+    float max_cost = max_cost_;
+    VTR_ASSERT(max_cost != std::numeric_limits<float>::min());
+    if (min_cost == std::numeric_limits<float>::max()) {
+        min_cost = max_cost;
+    }
+    auto min_bucket = cost_to_int(min_cost);
+    auto max_bucket = cost_to_int(max_cost);
+
+    if (min_bucket < 0 || max_bucket < 0 || max_bucket > 1000000) {
+        // If min and max are close to each other, assume 3 orders of
```

Comment:
This comment doesn't seem to exactly line up with the code.  Please check the math (it doesn't seem to do what the comment says).

Reply:
The comment is valid?  Can you provide a counter example?